std/machines/hash/poseidon2_bb.asm

```diff
@@ -0,0 +1,212 @@
+use std::array;
+use std::check::assert;
+use std::utils::unchanged_until;
+use std::utils::force_bool;
+use std::utils::sum;
+use std::convert::expr;
+use std::machines::memory::Memory;
+use std::machines::split::split_bb::SplitBB;
+
+// Implements the Poseidon2 permutation for the BabyBear.
+//
+// Apparently it can be used to hash arbitrary sized data by using the
+// Merkle–Damgård construction, or it can be used as a compression function
+// for building a Merkle tree.
+//
+// As it stands, it can not be used in a Sponge construction, because we don't
+// output the entire state.
+machine PoseidonGLMemory(mem: Memory, split_BB: SplitBB) with
+    latch: CKL_0,
+    operation_id: operation_id,
+    // Allow this machine to be connected via a permutation
+    call_selectors: sel,
+{
+    // The input data is passed via a memory pointer: the machine will read STATE_SIZE
+    // field elements from memory, in pairs of 16-bit limbs for BabyBear.
+    //
+    // Similarly, the output data is written to memory at the provided pointer.
+    //
+    // Reads happen at the provided time step; writes happen at the next time step.
+    operation poseidon_permutation<0> input_addr, output_addr, time_step ->;
+    col witness operation_id;
+
+    // Poseidon2 parameters, compatible with our powdr-plonky3 implementation.
+    //
+    // The the number of rounds to get 128-bit security was taken from here:
+    // https://github.com/Plonky3/Plonky3/blob/2df15fd05e2181b31b39525361aef0213fc76144/poseidon2/src/round_numbers.rs#L42
+
+    // S-box degree (this constant is actually not used, because we have to break the exponentiation into steps of at most degree 3).
+    let SBOX_DEGREE: int = 7;
+
+    // Number of field elements in the state
+    let STATE_SIZE: int = 16;
+
+    // Number of output elements
+    // (TODO: to use the Sponge construction, the entire state should be output)
+    let OUTPUT_SIZE: int = 8;
+
+    // Half the number of external rounds (half of external rounds happen before and half after the internal rounds).
+    let HALF_EXTERNAL_ROUNDS: int = 4;
+
+    // Number of internal rounds
+    let INTERNAL_ROUNDS: int = 13;
+
+    // External round MDS matrix
+    let MDS = [
+        [4, 6, 2, 2, 2, 3, 1, 1, 2, 3, 1, 1, 2, 3, 1, 1],
```

Comment:
Could we please avoid copying these magic constants across 4-5 files?

Reply:
They are not. The only other place most of these constants appear is inside plonky3 rust code.

The best I can do is, after merging this branch with the poseidon_bb one, place a fraction of the big list of random numbers that is common between the two of them in a separate file, as they are the output of the same deterministic RNG.